be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -2,80 +2,158 @@
 
 #include "exec/pipeline/pipeline_driver_queue.h"
 
+#include "exec/workgroup/work_group.h"
 #include "gutil/strings/substitute.h"
+
 namespace starrocks::pipeline {
-void QuerySharedDriverQueue::close() {
+
+void QuerySharedDriverQueue::put_back(const DriverRawPtr driver, bool from_dispatcher) {
+    _put_back(driver, from_dispatcher);
+}
+
+void QuerySharedDriverQueue::put_back(const std::vector<DriverRawPtr>& drivers, bool from_dispatcher) {
+    for (auto driver : drivers) {
+        _put_back(driver, from_dispatcher);
+    }
+}
+
+StatusOr<DriverRawPtr> QuerySharedDriverQueue::take() {
+    // -1 means no candidates; else has candidate.
+    int queue_idx = -1;
+    double target_accu_time = 0;
+    DriverRawPtr driver_ptr;
+
+    for (int i = 0; i < QUEUE_SIZE; ++i) {
+        // we just search for queue has element
+        if (!_queues[i].queue.empty()) {
+            double local_target_time = _queues[i].accu_time_after_divisor();
+            // if this is first queue that has element, we select it;
+            // else we choose queue that the execution time is less sufficient,
+            // and record time.
+            if (queue_idx < 0 || local_target_time < target_accu_time) {
+                target_accu_time = local_target_time;
+                queue_idx = i;
+            }
+        }
+    }
+
+    DCHECK(queue_idx >= 0);
```

Comment:
always return a not nullptr?

Reply:
Yes. `DriverQueueWithWorkGroup::take()` calls `QuerySharedDriverQueue::take()`, only when there is at least one driver in `QuerySharedDriverQueue`. Let me add some comment.